lib/inc/pxp-agent/util/bolt_module.hpp

```diff
@@ -0,0 +1,72 @@
+#ifndef SRC_UTIL_BOLT_MODULE_HPP_
+#define SRC_UTIL_BOLT_MODULE_HPP_
+
+#include <pxp-agent/action_response.hpp>
+#include <pxp-agent/module.hpp>
+#include <pxp-agent/results_storage.hpp>
+
+#include <leatherman/execution/execution.hpp>
+
+#include <boost/filesystem.hpp>
+
+namespace PXPAgent {
+namespace Util {
+
+// CommandObject holds collected parameters for leatherman's execution methods
+struct CommandObject {
+    std::string executable;
+    std::vector<std::string> arguments;
+    std::map<std::string, std::string> environment;
+    std::string input;
+    std::function<void(size_t)> pid_callback;
+};
+
+// This module is a basis for PXP modules supporting bolt functionality
+class BoltModule : public PXPAgent::Module {
+    public:
+        explicit BoltModule(std::shared_ptr<ResultsStorage> storage)
+           : storage_(std::move(storage)) {}
+
+        /// Whether the module supports non-blocking / asynchronous requests.
+        bool supportsAsync() override { return true; }
+
+        /// Log information about the output of the performed action
+        /// while validating the output is valid UTF-8.
+        /// Update the metadata of the ActionResponse instance (the
+        /// 'results_are_valid', 'status', and 'execution_error' entries
+        /// will be set appropriately; 'end' will be set to the current
+        /// time).
+        /// This function does not throw a ProcessingError in case of
+        /// invalid output on stdout; such failure is instead reported
+        /// in the response object's metadata.
+        void processOutputAndUpdateMetadata(ActionResponse& response) override;
```

Comment:
Should this be split? Having `and` in a method name is usually code smell, I'm curious whether it'd make sense to separate the two.

Reply:
I do see that this particular change is just moving existing code around, so perhaps not necessary to address this here.

Reply:
that makes sense though -- I'll just add a new commit to do it

Reply:
On second thought, this is an overridden method from the base Bolt module -- it doesn't seem wise to do that sort of refactor as a quick change in this PR

Reply:
I don't think the function needs to be refactored. It really just parses output then calls other functions based on that parsing.

I also think since it's just moving things around we can leave the name for now.